File: algorithms/Reaching Points/Reaching Points.py
from collections import defaultdict

class Solution:
    def reachingPoints(self, sx: int, sy: int, tx: int, ty: int) -> bool:
        def nextNum(ta,tb,s):
            if ta % tb == s % tb:
                return min(ta, s)
            return ta % tb
        visited = defaultdict(bool)
        while tx >= sx and ty >= sy and (sx != tx or sy != ty):
            if tx > ty:
                tx, ty = nextNum(tx, ty, sx), ty
            else:
                tx, ty = tx, nextNum(ty, tx, sy)
            if visited[(tx,ty)]:
                break
            visited[(tx,ty)] = True
        return (sx == tx) and (sy == ty)

File: scripts/algorithms/R/Reaching Points/Reaching Points.java
// Runtime: 1 ms (Top 55.54%) | Memory: 41.4 MB (Top 26.41%)
class Solution {
    public boolean reachingPoints(int sx, int sy, int tx, int ty) {
        if (sx==tx && sy==ty){
            return true;
        }
        if (tx < sx || ty < sy){
            return false;
        }
        return tx<ty?
                  reachingPoints(sx, sy, tx, Math.min(ty%tx+sy/tx*tx,ty-tx))
                : reachingPoints(sx, sy, Math.min(tx%ty+sx/ty*ty,tx-ty), ty);
    }
}